server/src/modules/maps/maps.service.ts

```diff
@@ -440,6 +449,226 @@ export class MapsService {
+            flags: dbResponse.flags,
+            trackNum: (dbResponse as any).run.trackNum,
+            zoneNum: (dbResponse as any).run.zoneNum,
+            rank: dbResponse.rank,
+            rankXP: dbResponse.rankXP,
+            createdAt: dbResponse.createdAt,
+            updatedAt: dbResponse.updatedAt
+        });
+    }
+
+    async getRankAround(userID: number, mapID: number, query: MapRankGetNumberQuery): Promise<MapRankDto[]> {
+        const where: Prisma.UserMapRankWhereInput = {
+            mapID: mapID,
+            flags: 0,
+            userID: userID,
+            run: {
+                trackNum: 0,
+                zoneNum: 0
+            }
+        };
+
+        if (query.flags) where.flags = query.flags;
+        if (query.trackNum) where.run.trackNum = query.trackNum;
+        if (query.zoneNum) where.run.zoneNum = query.zoneNum;
+
+        const include = {
+            run: true
+        };
+
+        const order: Prisma.UserMapRankOrderByWithAggregationInput = {
+            rank: 'asc'
+        };
+
+        const myRankInfo = await this.mapRepo.getRank(where, include);
+
+        if (!myRankInfo) throw new NotFoundException('No personal best detected');
+
+        const myRank = myRankInfo.rank;
+
+        where.userID = undefined;
+
+        const ranks = await this.mapRepo.getRanks(where, include, order, Math.max(myRank - 6, 0), 11);
+
+        //This is done because the MapRankDto still contains trackNum and zoneNum but UserMapRank doesn't
+        const ret = ranks[0].map((item) =>
+            DtoFactory(MapRankDto, {
+                mapID: item.mapID,
+                userID: item.userID,
+                runID: item.runID,
+                gameType: item.gameType,
+                flags: item.flags,
+                trackNum: (item as any).run.trackNum,
+                zoneNum: (item as any).run.zoneNum,
+                rank: item.rank,
+                rankXP: item.rankXP,
+                createdAt: item.createdAt,
+                updatedAt: item.updatedAt
+            })
+        );
+
+        return ret;
+    }
+
+    async getRankFriends(userID: number, mapID: number, query: MapRankGetNumberQuery): Promise<MapRankDto[]> {
+        const user = await this.userRepo.get(userID);
+        const steamID = user.steamID;
+
+        const res = await firstValueFrom(
+            this.http
+                .get('https://api.steampowered.com/ISteamUser/GetFriendList/v1/', {
+                    params: {
+                        key: this.config.get('steam.webAPIKey'),
+                        steamID: steamID,
+                        relationship: 'friend'
+                    }
+                })
+                .pipe(
+                    catchError((error: AxiosError) => {
+                        if (error.response.status === 401)
+                            throw new ConflictException('Friends list or profile is private');
+
+                        throw new BadRequestException('Bad request');
```

Comment:
If we don't get a response from Steam, that's really a problem on our end, not the client. Remember, 4xxs are *client* errors such as bad request data or attempting to access a resource that doesn't exist.

We're responsible for the services our system uses, including Steam, so if there's something wrong in their end (classic Tuesday downtime, thanks Valve), so lets throw a `ServiceUnavailableException('Failed to retrieve friends list from Steam');` or whatever Nest calls its 503 exceptions.